src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/TextEncoder.cs

```diff
@@ -693,46 +700,130 @@ private unsafe int FindFirstCharacterToEncode(ReadOnlySpan<char> text)
             // input sequence. If we consume the entire text without seeing either of these, return -1 to indicate
             // that the text can be copied as-is without escaping.
 
-            int i = 0;
-            while (i < utf8Text.Length)
+            fixed (byte* ptr = utf8Text)
             {
-                byte value = utf8Text[i];
-                if (UnicodeUtility.IsAsciiCodePoint(value))
+                int idx = 0;
+
+#if NETCOREAPP
+                if (Sse2.IsSupported)
                 {
-                    if (!ReferenceEquals(GetAsciiEncoding(value), s_noEscape))
+                    sbyte* startingAddress = (sbyte*)ptr;
+                    while (utf8Text.Length - 16 >= idx)
                     {
-                        return originalUtf8TextLength - utf8Text.Length + i;
+                        Debug.Assert(startingAddress >= ptr && startingAddress <= (ptr + utf8Text.Length - 16));
+
+                        // Load the next 16 bytes.
+                        Vector128<sbyte> sourceValue = Sse2.LoadVector128(startingAddress);
+
+                        Vector128<sbyte> mask = Sse2Helper.CreateAsciiMask(sourceValue);
+                        int index = Sse2.MoveMask(mask.AsByte());
+
+                        int processNextSixteen = idx + 16;
+                        Debug.Assert(processNextSixteen <= utf8Text.Length);
+                        if (index != 0)
+                        {
+                            // At least one of the following 16 bytes is non-ASCII.
+                            while (idx < processNextSixteen)
+                            {
+                                Debug.Assert((ptr + idx) <= (ptr + utf8Text.Length));
+
+                                if (UnicodeUtility.IsAsciiCodePoint(ptr[idx]))
+                                {
+                                    if (DoesAsciiNeedEncoding(ptr[idx]) == 1)
+                                    {
+                                        goto Return;
+                                    }
+                                    idx++;
+                                }
+                                else
+                                {
+                                    OperationStatus opStatus = UnicodeHelpers.DecodeScalarValueFromUtf8(utf8Text.Slice(idx), out uint nextScalarValue, out int utf8BytesConsumedForScalar);
+
+                                    Debug.Assert(nextScalarValue <= int.MaxValue);
+                                    if (opStatus != OperationStatus.Done || WillEncode((int)nextScalarValue))
+                                    {
+                                        goto Return;
+                                    }
+
+                                    Debug.Assert(opStatus == OperationStatus.Done);
+                                    idx += utf8BytesConsumedForScalar;
+                                }
+                            }
```

Comment:
The same code is used in the sequential path. Refactor to a method?

Reply:
I have in-line gotos here making it hard to refactor out into a method. Let me try and see if perf is impacted.

Reply:
There's a small regression when I do this (~10%, goes from 12.5 ns to 14 ns, dealing with 7 characters).

Here's basically what I changed:
```C#
while (idx < utf8Text.Length)
{
    if (NeedsEscapingHelper(ptr, ref idx, utf8Text))
    {
        goto Return;
    }
}

[MethodImpl(MethodImplOptions.AggressiveInlining)]
private unsafe bool NeedsEscapingHelper(byte* ptr, ref int idx, ReadOnlySpan<byte> utf8Text)
{
    Debug.Assert((ptr + idx) <= (ptr + utf8Text.Length));

    bool result = true;

    if (UnicodeUtility.IsAsciiCodePoint(ptr[idx]))
    {
        if (DoesAsciiNeedEncoding(ptr[idx]) == 1)
        {
            goto Return;
        }
        idx++;
    }
    else
    {
        OperationStatus opStatus = UnicodeHelpers.DecodeScalarValueFromUtf8(utf8Text.Slice(idx), out uint nextScalarValue, out int utf8BytesConsumedForScalar);

        Debug.Assert(nextScalarValue <= int.MaxValue);
        if (opStatus != OperationStatus.Done || WillEncode((int)nextScalarValue))
        {
            goto Return;
        }

        Debug.Assert(opStatus == OperationStatus.Done);
        idx += utf8BytesConsumedForScalar;
    }

    result = false;

Return:
    return result;
}
```